src/wifi.rs

```diff
@@ -1442,3 +1446,141 @@ impl WifiWait {
         }
     }
 }
+
+pub struct AsyncWifiDriver<'d, T>
+where
+    T: BorrowMut<WifiDriver<'d>>,
+{
+    base_driver: T,
+    _p: PhantomData<WifiDriver<'d>>,
+}
+
+impl<'d, T> AsyncWifiDriver<'d, T>
+where
+    T: BorrowMut<WifiDriver<'d>>,
+{
+    #[cfg(all(feature = "alloc", esp_idf_comp_nvs_flash_enabled))]
+    pub fn new(base_driver: T) -> Result<Self, EspError> {
+        Ok(AsyncWifiDriver {
+            base_driver,
+            _p: PhantomData,
+        })
+    }
+
+    pub fn get_capabilities(&self) -> Result<EnumSet<Capability>, EspError> {
+        self.base_driver.borrow().get_capabilities()
+    }
+
+    pub fn get_configuration(&self) -> Result<Configuration, EspError> {
+        self.base_driver.borrow().get_configuration()
+    }
+
+    pub fn set_configuration(&mut self, conf: &Configuration) -> Result<(), EspError> {
+        self.base_driver.borrow_mut().set_configuration(conf)
+    }
+
+    pub fn is_started(&self) -> Result<bool, EspError> {
+        self.base_driver.borrow().is_started()
+    }
+
+    pub fn is_connected(&self) -> Result<bool, EspError> {
+        self.base_driver.borrow().is_connected()
+    }
+
+    pub async fn start(&mut self) -> Result<(), EspError> {
+        self.base_driver.borrow_mut().start()
+        // TODO: wait until start is complete
+    }
+
+    pub async fn stop(&mut self) -> Result<(), EspError> {
+        self.base_driver.borrow_mut().stop()
+        // TODO: wait until stop is complete?
+    }
+
+    pub async fn connect(&mut self) -> Result<(), EspError> {
+        self.base_driver.borrow_mut().connect()
+        // TODO: wait until connect is complete
```

Comment:
FYI if you would like a pointer on how you might implement this, you can take a look at what I did for the bare metal esp-wifi: https://github.com/esp-rs/esp-wifi/blob/76ba312b55b841cbbe9090c4f8dfb12ad379bb38/esp-wifi/src/wifi/mod.rs#L1437-L1453

Reply:
Once you get one event working, it should be straight forward for the rest

Reply:
Thanks, I'll have a look!